src/toxcore/net_crypto/net_crypto.rs

```diff
@@ -0,0 +1,219 @@
+        let connections = self.connections.read();
+        let keys_by_addr = self.keys_by_addr.read();
+        let connection = keys_by_addr.get(&(addr.ip(), addr.port())).cloned().and_then(|pk|
+            connections.get(&pk)
+        );
+        if let Some(connection) = connection {
+            let mut connection = connection.write();
+            connection.update_udp_received_time();
+            Box::new(future::result(
+                connection.handle_crypto_handshake(packet, &self.symmetric_key)
+            ))
+        } else {
+            Box::new(future::err( // TODO: create crypto connection
+                Error::new(
+                    ErrorKind::Other,
+                    format!("No crypto connection for address {}", addr)
+                )
+            ))
+        }
+    }
+
+    /// Send packet to crypto connection choosing TCP or UDP protocol
+    fn send_packet(&self, packet: DhtPacket, connection: &mut CryptoConnection) -> IoFuture<()> {
+        if let Some(addr) = connection.udp_addr {
+            if connection.is_udp_alive() {
+                return self.send_to_udp(addr, packet)
+            }
+
+            let udp_attempt_should_be_made = connection.udp_attempt_should_be_made() && {
+                // check if the packet is not too big
+                let mut buf = [0; 95];
+                packet.to_bytes((&mut buf, 0)).is_ok()
+            };
+
+            if udp_attempt_should_be_made {
+                connection.update_udp_send_attempt_time();
+                self.send_to_udp(addr, packet)
+            } else {
+                Box::new(future::ok(()))
+            }
+        } else {
+            Box::new(future::ok(()))
+        }
+
+        // TODO: send via TCP relay here
+    }
+
+    /// Send `CookieRequest` or `CryptoHandshake` packet if needed depending on
+    /// connection status and update sent counter
+    fn send_status_packet(&self, connection: &mut CryptoConnection) -> IoFuture<()> {
+        match connection.packet_to_send() {
+            Some(packet) => self.send_packet(packet, connection), // TODO: collect packets and send via send_all
+            None => Box::new(future::ok(())),
+        }
+    }
+
+    /// The main loop that should be run at least 20 times per second
+    pub fn main_loop(&self) -> IoFuture<()> {
+        let connections = self.connections.read();
+        let mut send_futures = Vec::new();
+        let mut timed_out = Vec::new();
+        // Only one cycle over all connections to prevent many lock acquirements
+        for (&pk, connection) in connections.iter() {
+            let mut connection = connection.write();
+
+            if connection.is_timed_out() {
+                timed_out.push((pk, connection.udp_addr));
+                continue;
+            }
+
+            let send_future = self.send_status_packet(connection.deref_mut());
+            send_futures.push(send_future);
```

Comment:
use connections.iter().map collect instead of mutable send_futures

Reply:
The only way to rewrite this cycle to map:
```rust
        let mut timed_out = Vec::new();
        // Only one cycle over all connections to prevent many lock acquirements
        let send_futures = connections.iter().map(|(&pk, connection)| {
            (pk, connection.write())
        }).filter(|&(pk, ref connection)| {
            if connection.is_timed_out() {
                timed_out.push((pk, connection.udp_addr));
                false
            } else {
                true
            }
        }).map(|(_, mut connection)| {
            self.send_status_packet(connection.deref_mut())
        }).collect::<Vec<_>>();
```
It has two `map`s and dirty `filter` with side effect. You sure that it worth it?